core/src/main/kotlin/fr/sncf/osrd/stdcm/preprocessing/implementation/BlockAvailability.kt

```diff
@@ -34,13 +36,153 @@ data class BlockAvailability(
+        var timeShift = 0.0
+        var firstConflictOffset = infraExplorer.getIncrementalPath().toTravelledPath(startOffset)
+        while (timeShift.isFinite()) {
+            val shiftedStartTime = startTime + timeShift
+            val pathStartTime =
+                shiftedStartTime - infraExplorer.interpolateDepartureFromClamp(startOffset)
+            val endTime = infraExplorer.interpolateDepartureFromClamp(endOffset) + pathStartTime
+
+            val stepAvailability =
+                getStepAvailability(
+                    infraExplorer,
+                    startOffset,
+                    endOffset,
+                    pathStartTime,
+                    shiftedStartTime
+                )
+            val conflictAvailability =
+                getConflictAvailability(
+                    infraExplorer,
+                    startOffset,
+                    pathStartTime,
+                    shiftedStartTime,
+                    endTime
+                )
+            val availability =
+                getMostRestrictiveAvailability(stepAvailability, conflictAvailability)
+
+            if (availability !is BlockAvailabilityInterface.Unavailable) {
+                if (timeShift > 0.0) { // Availability is available due to adding a delay: timeShift
+                    return BlockAvailabilityInterface.Unavailable(timeShift, firstConflictOffset)
+                }
+                // Availability is directly available without adding any delay
+                return availability
+            }
+            timeShift += availability.duration
+            firstConflictOffset = availability.firstConflictOffset
+        }
+        // No available solution with a finite delay was found
+        return BlockAvailabilityInterface.Unavailable(Double.POSITIVE_INFINITY, firstConflictOffset)
+    }
+
+    /**
+     * Check that the planned step timing data is respected, and return the corresponding
+     * availability.
+     * - If some steps are not respected, find the minimumDelayToBecomeAvailable and return
+     *   Unavailable(minimumDelayToBecomeAvailable, ...). In case this delay fails other steps,
+     *   return Unavailable(Infinity, ...).
+     * - If all steps are respected, find the maximumDelayToStayAvailable and return
+     *   Available(maximumDelayToStayAvailable, ...).
+     */
+    private fun getStepAvailability(
+        infraExplorer: InfraExplorerWithEnvelope,
+        startOffset: Offset<Path>,
+        endOffset: Offset<Path>,
+        pathStartTime: Double,
+        startTime: Double,
+    ): BlockAvailabilityInterface.Availability {
+        var maximumDelayToStayAvailable = Double.POSITIVE_INFINITY
+        var minimumDelayToBecomeAvailable = 0.0
+        var firstConflictOffset = Offset<TravelledPath>(Double.POSITIVE_INFINITY.meters)
+
+        val incrementalPath = infraExplorer.getIncrementalPath()
+        first@ for (step in plannedSteps) {
+            for (location in step.locations) {
```

Comment:
Moving the inner loop into its own method would remove the need for a label

Generally speaking the method could be broken down into smaller ones